### PR TITLE
Convert vSphere resource names to full paths

### PR DESCRIPTION
**Pull request by saimanoj01**
### What this PR does / why we need it
Right now the user can pass just the vSphere resource names (eg. VSPHERE_FORLDER: "vm0") when creating TKG clusters on vSphere. This is causing problems for users with complex(huge) vSphere setups. This change will convert those names to fully qualified paths.

### Which issue(s) this PR fixes

Fixes https://github.com/vmware-tanzu/tanzu-framework/issues/1356

### Describe testing done for PR
Used the below config to create a workload cluster on vSphere.
```
VSPHERE_DATACENTER: dc0
VSPHERE_DATASTORE: sharedVmfs-1
VSPHERE_FOLDER: "vm"
VSPHERE_NETWORK: "VM Network"
VSPHERE_RESOURCE_POOL: cluster0/Resources
```

The above resources are converted to full paths when generating the templates.
```
Setting config variable "VSPHERE_DATACENTER" to value "/dc0"
Setting config variable "VSPHERE_NETWORK" to value "/dc0/network/VM Network"
Setting config variable "VSPHERE_RESOURCE_POOL" to value "/dc0/host/cluster0/Resources"
Setting config variable "VSPHERE_DATASTORE" to value "/dc0/datastore/sharedVmfs-1"
Setting config variable "VSPHERE_FOLDER" to value "/dc0/vm"
```

### Release note

```release-note
Convert resource names to full paths when creating clusters on vSphere
```

### PR Checklist

- [ ] Squash the commits into one or a small number of logical commits
      
- [ ] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [ ] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: pkg/v1/tkg/client/validate.go [hunk 1]
```diff
@@ -813,39 +813,73 @@ func (c *TkgClient) ValidateVsphereVipWorkloadCluster(clusterClient clusterclien
...
+
 	for _, resourceType := range VsphereResourceType {
-		path, err := c.TKGConfigReaderWriter().Get(resourceType)
+		path, err = c.TKGConfigReaderWriter().Get(resourceType)
 		if err != nil {
 			continue
 		}
 
 		switch resourceType {
+		case constants.ConfigVariableVsphereDatacenter:
```
Comment (vuil) [inline pkg/v1/tkg/client/validate.go:826]: What is the finder's behavior when multiple resources of the same name and type is present in the vSphere inventory? 
The one situation we should definitely avoid is arbitrarily selecting one of a viable fully qualified paths in that case. 
Also a good way to explain the code's behavior is to set up the unit tests to demonstrate how such ambiguities are being handled.

File: pkg/v1/tkg/client/validate.go [hunk 2]
```diff
@@ -813,39 +813,73 @@ func (c *TkgClient) ValidateVsphereVipWorkloadCluster(clusterClient clusterclien
 
 // ValidateVsphereResources validates vsphere resource path specified in tkgconfig
 func (c *TkgClient) ValidateVsphereResources(vcClient vc.Client, dcPath string) error {
+	var path, resourceMoid string
+	var err error
+
 	for _, resourceType := range VsphereResourceType {
```
Comment (vuil) [inline pkg/v1/tkg/client/validate.go:819]: nit, preexisting: VsphereResourceType -> VsphereResourceTypes or VsphereResourceConfigKeys?

Comment (saimanoj01) [reply pkg/v1/tkg/client/validate.go]: Done. Thanks for the suggestion.

File: pkg/v1/tkg/client/validate.go (hunk 1 above)
Comment (saimanoj01) [reply pkg/v1/tkg/client/validate.go:827]: Added unit tests.

**Comment by saimanoj01** [comment]
Created an issue to improve ux when duplicate resources with the same name are present.
https://github.com/vmware-tanzu/tanzu-framework/issues/1364

**Review by vuil — APPROVED** [review]
Thanks for the updates. lgtm!

**Pull request merged**